mux: implement ipc4 support

fkwasowi opened the pull request:
mux module does not support ipc4

A change is needed to work properly: https://github.com/thesofproject/rimage/pull/112

lgirdwood reviewed (commented):
Lots of complex looking code in here, can you provide more inline comments on each block to help reviewers. Thanks.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -34,11 +34,15 @@
 static const struct comp_driver comp_mux;
```
In commit, add space after prefix "mux: implement ipc4 support"

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	cd->config.frame_format_deprecated = cd->md.base_cfg.audio_fmt.depth;
+	cd->config.num_channels_deprecated = cd->md.base_cfg.audio_fmt.channels_count;
+	cd->config.num_streams = 2;
+	/* clear masks */
```
Adding empty line will increase readability.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	for (uint16_t i = 0; i < cd->config.num_streams; i++)
+		for (uint16_t j = 0 ; j < PLATFORM_MAX_CHANNELS; j++)
+			cd->config.streams[i].mask[j] = 0;
+	int mask = 1;
```
Declare variable at start of scope.

marc-hb replied:
This is not required since C99 (or any other modern programming language) and it restricts the use of `const`.

We had places where declarations are not at the top of the block (#3995) and they never caused any issue with any compiler.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+				cd->md.base_cfg.audio_fmt.channels_count;
+		mask = mask * 2;
+	}
+	for (uint16_t i = 0; i < cd->md.reference_format.channels_count; i++) {
```
Empty line would help here as well.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+			}
+		}
+	}
+	if (mux_mix_check(&cd->config)) {
```
Ditto

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+		return NULL;
+	}
+	dev->state = COMP_STATE_READY;
+	return dev;
```
Ditto.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	/* set masks for all streams */
+	for (uint16_t i = 0; i < cd->md.base_cfg.audio_fmt.channels_count; i++) {
+		cd->config.streams[0].mask[i] = mask;
+		cd->config.streams[0].num_channels_deprecated =
```
Whye are we using "deprecated" field?

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+static void set_mux_params(struct comp_dev *dev,
```
Empty line before function definition.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+		buffer_release(sink_c);
+	}
+
+
```
Unnecessary empty line.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+					__sparse_cache &source_c->stream);
+			j = source_c->id;
+			cd->config.streams[j].pipeline_id = source_c->pipeline_id;
+			if (j == 0) {
```
Comment would help, why does "j == 0" get special treatment?

fkwasowi replied:
struct mux_data {
	struct ipc4_base_module_cfg base_cfg;
	struct ipc4_audio_format reference_format;
	struct ipc4_audio_format output_format;
};
In the mux module we have two inputs and one output, for the first input we assign parameters from base cfg, for the second from reference_format, and for sink output_format

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	return 0;
+}
+#endif
 static void mux_free(struct comp_dev *dev)
```
Add empty line after L395.

lyakh replied:
yes, an empty line after `#endif` would be good

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -711,13 +905,22 @@ static int mux_trigger(struct comp_dev *dev, int cmd)
-	case COMP_TRIGGER_PRE_START:
-		if (src_n_active || src_n_paused)
-			return PPL_STATUS_PATH_STOP;
+	if (dir == SOF_IPC_STREAM_PLAYBACK) {
```
Isn't this changing the functionality for IPC3 as well as this code is only entered if direction is playback?

marc-hb replied:
I don't know all the details but it is strange to this question "resolved" without any answer.

Please resolve/hide only comments where it is obvious for everyone that all parties have come to 100% agreement and understanding. Even then it's sometimes good to leave some interesting discussions and rationales visible. Of course hide stuff like "missing new line".

There is no policy in this project to resolve all comments before merging. It is often useful to read some code review discussions even after the code is merged and resolved comments get in the way of that.

kv2019i reviewed (changes requested):
Thanks @fkwasowi ! A few comments inline, please see.

juimonen commented:
@fkwasowi is it so that copieror mixinmixout can't do the operations the mux/demux is doing? Anyway good to have this, but we need to get the "full story" how these parameters are coming from the user space via topology and kernel to firmware. As I understand current mux/demux is a "process component" with binary blob interface -> so it should be ipc agnostic? so why do we need a new interface for it?

aiChaoSONG commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,217 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	 * reference. For example, for base config 2-channel and reference 2-channel, masks
+	 * should look like: mask[] = { 0b00000001, 0b00000010 } for the first
+	 * stream and mask[] = { 0b00000100, 0b00001000 } for the second
+	 * (reference). This means that the first channel of first stream
+	 * will be copied to the first output channel, second channel of the first stream
+	 * will be copied to the second output channel, first channel of second stream
+	 * will be copied to the third output channel, the second channel of second stream
+	 * will be copied to the fourth output channel.
```
A small graph explains better:
```
          +---+           +---+
          | 0 |---------> | 0 |
INPUT     +---+           +---+ 
STREAM 0  | 1 |---------> | 1 |
          +---+           +---+  OUTPUT
                   +----> | 2 |  STREAM
          +---+    |      +---+
          | 0 |----+  +-> | 3 |
INPUT     +---+       |   +---+
STREAM 1  | 1 |-------+
          +---+

```

aiChaoSONG commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,217 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	for (uint16_t i = 0; i < cd->config.num_streams; i++) {
+		for (uint16_t j = 0 ; j < PLATFORM_MAX_CHANNELS; j++) {
+			if (popcount(cd->config.streams[i].mask[j]) > 1) {
+				comp_cl_err(&comp_mux, "build_config(): mux component is not able to mix channels");
```
I think this error message is confusing, maybe `build_config(): invalid mask for stream %d, channel:%d`?

aiChaoSONG commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +204,217 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	ret = build_config(cd, dev);
+	assert(!ret);
+
+	if (ret < 0) {
+		rfree(cd);
+		rfree(dev);
+		return NULL;
+	}
```
this code is not reachable as you have used assert on ret.

kv2019i reviewed (changes requested):
Thanks, the comment in build_config() is very good. Only one very minor nit left, please see inline, otherwise seems good to go.

kv2019i commented:
@fkwasowi For the rimage dependency, please add a separate commit to this PR that updates sof/west.yml to point to the rimage repository where the dependent patch is merged. Please contact me offline if you need help. Other option is to do a separate PR to update the west.yml, and merge that first and then this one. Both are good to me.

lyakh commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,230 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	}
+
+	for (uint16_t i = 0; i < cd->md.reference_format.channels_count; i++) {
+		cd->config.streams[1].mask[i] = mask;
```
I think it is correct as your comment above explains, but just to double check: `mask` continues from above, right? You don't need to reset it to 1 before entering the loop?

fkwasowi replied:
The mask is reset above (line 220)

lyakh replied:
yes, I meant you don't reset it between these two loops - for streams 0 and 1.

lyakh commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,230 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	/* verification that masks are correct */
+	for (uint16_t i = 0; i < cd->config.num_streams; i++) {
+		for (uint16_t j = 0 ; j < PLATFORM_MAX_CHANNELS; j++) {
+			if (popcount(cd->config.streams[i].mask[j]) > 1) {
```
wait, you just calculated them, right? Isn't it similar to
```
x = 2 + 2;
if (x != 4) return -ECPUMELTDOWN;
```
? Or can there be more than 2 streams so that the second and any further streams weren't set up above? And you probably want to `memset(0)` the whole array at the top of the function if it's important that skipped masks are valid too?

lyakh replied:
but how can any of those masks contain more than 1 set bit? You begin setting them with `mask = 1` and then just shift the mask to the left, so each mask, that has been set, will have exactly one bit set.

fkwasowi replied:
I removed the redundant verification and added memset

lyakh commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,230 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+		sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		sink_c = buffer_acquire(sink);
+		audio_stream_init_alignment_constants(byte_align, frame_align_req,
+						      __sparse_cache &sink_c->stream);
```
`__sparse_cache` shouldn't be needed here

lyakh commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,230 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+			source = container_of(source_list, struct comp_buffer, sink_list);
+			source_c = buffer_acquire(source);
+			audio_stream_init_alignment_constants(byte_align, frame_align_req,
+							      __sparse_cache &source_c->stream);
```
ditto

fkwasowi commented:
> @fkwasowi is it so that copier or mixinmixout can't do the operations the mux/demux is doing? Anyway good to have this, but we need to get the "full story" how these parameters are coming from the user space via topology and kernel to firmware. As I understand current mux/demux is a "process component" with binary blob interface -> so it should be ipc agnostic? so why do we need a new interface for it?

Modul copier and mixin/mixout at the moment do not have the ability to operate in the same way as Mux. It is possible that in the future the copier will have such functionality, but at the moment it does not. IPC3 and IPC4 significantly differ in config, so there was a need to redesign the module to support IPC4.

fkwasowi commented:
> @fkwasowi For the rimage dependency, please add a separate commit to this PR that updates sof/west.yml to point to the rimage repository where the dependent patch is merged. Please contact me offline if you need help. Other option is to do a separate PR to update the west.yml, and merge that first and then this one. Both are good to me.

A separate PR has been created: https://github.com/thesofproject/rimage/pull/110, which is already merged

kv2019i commented:
@fkwasowi wrote:
> A separate PR has been created: [thesofproject/rimage#110](https://github.com/thesofproject/rimage/pull/110), which is already merged

The rimage changes are not automatically visible in SOF builds. Rather if a newer rimage (or Zephyr) is needed, one needs to submit a SOF PR to update sof/west.yml and change the rimage (or Zephyr) point to a more recent commit-id. All our CI infra uses west.yml to pick which versions to use. I.e. you need to add a separate commit to this PR to update rimage in west.yml, or otherwise the IPC4 tests will fail.

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,231 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+			cd->config.streams[j].pipeline_id = source_c->pipeline_id;
+			valid_bit_depth = cd->md.base_cfg.audio_fmt.valid_bit_depth;
+			/* set parameters for speech input channels */
+			if (j == BASEFW_QUEUE_ID) {
```
Sorry my bad, I guess this has nothing to do with BASEFW, so better to rename this to BASE_CFG_QUEUED_ID .

And you still have the confusing comment about "speech input channels", please address this (either expand why this is related to speech, or replace with something understandable).

kv2019i commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,231 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+static int build_config(struct comp_data *cd, struct comp_dev *dev)
+{
+	dev->ipc_config.type = SOF_COMP_MUX;
+	cd->config.frame_format_deprecated = cd->md.base_cfg.audio_fmt.depth;
```
One use of setting "frame_format_deprecated" still. This is not needed. This field has not been used since 23ae5d824ac180232835df355403a3a838b02fcc commit in 2020.

kv2019i reviewed (changes requested):
Starts to be ready to go. You do need to add an update to west.yml in a seprate commti (see https://github.com/thesofproject/sof/pull/6396 for an example) and have that as a separate git commit in this PR. A couple of minor thing, including one bad naming I proposed (sorry about that).

fkwasowi commented:
> Starts to be ready to go. You do need to add an update to west.yml in a seprate commti (see #6396 for an example) and have that as a separate git commit in this PR. A couple of minor thing, including one bad naming I proposed (sorry about that).

Done (I have added a separate commit to west.yml)

kv2019i reviewed (approved):
Thanks @fkwasowi , looks good now. FYI @marc-hb , this PR updates rimage as well.

fkwasowi commented on west.yml:
```diff
@@ -16,7 +16,7 @@ manifest:
       repo-path: rimage
       path: sof/rimage
-      revision: 542924d70c1715671ad8213440f01dc6dadb52e4
+      revision: f1e5621067bb530028c00d04a3abe19911285f1d
```
Ok, first wait for #6398 to be merged and then I can already upload my change?

marc-hb replied:
Yes. @fkwasowi #6398 was just merged, please git fetch, rebase and follow the new instructions inside the west.yml file.

Make sure your west.yml change and your git submodule change are in the same, isolated commit and that git status and git describe --dirty are clean

Thanks for your patience!

mmaka1 commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,229 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+		for (j = 0 ; j < PLATFORM_MAX_CHANNELS; j++)
+			cd->config.streams[i].mask[j] = 0;
+	/*
+	 * Setting masks for streams. Channels for IPC4 are mapped
```
Any documentation of the IPC messages should be maintained in the headers and written in doxygen (javadoc) format. The IPC interface specification will be generated from the headers and included in the sof-docs.

mmaka1 commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,229 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	int i, j;
+
+	/* clear masks */
+	for (i = 0; i < cd->config.num_streams; i++)
```
Is that verified somewhere that no OOB occurs when iterating over cd->config.streams[] using num_streams as the limit?

fkwasowi replied:
I do not think so, because the copy function takes into account only those streams/channels on which the mask is set, and this function is performed only for ipc4, where we have only two streams. The change has been tested.

mmaka1 replied:
Still there is no clear relation between the limit of the loop and the allocation size in this function, code looks unsafe.

fkwasowi commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,229 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+static int build_config(struct comp_data *cd, struct comp_dev *dev)
+{
+	dev->ipc_config.type = SOF_COMP_MUX;
+	cd->config.num_streams = 2;
```
Because MUX_MAX_STREAMS for IPC3 is 4 , and in IPC4 we have 2. I can do the differentiation under ifdef.

fkwasowi commented on src/audio/mux/mux.c:
```diff
@@ -200,7 +205,229 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+	/* clear masks */
+	for (i = 0; i < cd->config.num_streams; i++)
+		for (j = 0 ; j < PLATFORM_MAX_CHANNELS; j++)
+			cd->config.streams[i].mask[j] = 0;
```
Done

marc-hb reviewed (dismissed):
Thanks for updating the rimage submodule as well, however there are still many serious rimage problems:

- In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, the  commit message _subject_ says "upgrade rimage to f1e5621067bb" but the commit message _body_ says "upgrade rimage to e9e074c3c3e8b7f". At least one of them is wrong (or both)

- In fact both are wrong because one of them is in `west.yml` but you upgraded the submodule to a third, different rimage commit d957e0368b86 instead!? As instructed in the new comments I added to `west.yml`, `git status` and `git describe --dirty` should report the submodule / west mismatch. Didn't they? Always look at "git status" before pushing (even when a single git repo is involved)

- The latest rimage commit is currently yet another rimage commit: 3ee717eebc6a , why not upgrade to the most recent rimage commit? Is there something wrong with it? Maybe because we all love dealing with rimage upgrades so much, save another rimage upgrade for later? :-D

- In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, your west.yml change is the LAST commit of the three. You're upgrading rimage  because your other commits need that rimage upgrade, correct? If correct then your rimage upgrade commit must be **first**. It cannot be last because every commit must be usable for `git bisect` and other test reasons.

- Wait, actually you upgraded the rimage git submodule and west.yml in... different commits! They must be kept in sync in every commit.

- You missed this log --oneline instruction below in the west.yml file. It's important because everyone regularly gets lost and confused with module updates - including you, see all the examples above. 

```
  # When upgrading projects here please run git log --oneline in the
  # project and if not too long then include the output in your commit
  # message. See examples of good commit messages at
  # https://github.com/thesofproject/sof/commits/4e1d3ba61abd3/rimage
  # (git log 4e1d3ba61abd3 -- rimage/)
```

Make sure you find a good git plugin in your editor (or a good standalone git graphical interface), one that lets you easily review all these changes before submitting. Never restrict yourself to the limited git command line, always use all the available git options; especially when ugprading a dependency.

fkwasowi commented:
> Thanks for updating the rimage submodule as well, however there are still many serious rimage problems:
> 
> * In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, the  commit message _subject_ says "upgrade rimage to f1e5621067bb" but the commit message _body_ says "upgrade rimage to e9e074c3c3e8b7f". At least one of them is wrong (or both)
> * In fact both are wrong because one of them is in `west.yml` but you upgraded the submodule to a third, different rimage commit d957e0368b86 instead!? As instructed in the new comments I added to `west.yml`, `git status` and `git describe --dirty` should report the submodule / west mismatch. Didn't they? Always look at "git status" before pushing (even when a single git repo is involved)
> * The latest rimage commit is currently yet another rimage commit: 3ee717eebc6a , why not upgrade to the most recent rimage commit? Is there something wrong with it? Maybe because we all love dealing with rimage upgrades so much, save another rimage upgrade for later? :-D
> * In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, your west.yml change is the LAST commit of the three. You're upgrading rimage  because your other commits need that rimage upgrade, correct? If correct then your rimage upgrade commit must be **first**. It cannot be last because every commit must be usable for `git bisect` and other test reasons.
> * Wait, actually you upgraded the rimage git submodule and west.yml in... different commits! They must be kept in sync in every commit.
> * You missed this log --oneline instruction below in the west.yml file. It's important because everyone regularly gets lost and confused with module updates - including you, see all the examples above.
> 
> ```
>   # When upgrading projects here please run git log --oneline in the
>   # project and if not too long then include the output in your commit
>   # message. See examples of good commit messages at
>   # https://github.com/thesofproject/sof/commits/4e1d3ba61abd3/rimage
>   # (git log 4e1d3ba61abd3 -- rimage/)
> ```
> 
> Make sure you find a good git plugin in your editor (or a good standalone git graphical interface), one that lets you easily review all these changes before submitting. Never restrict yourself to the limited git command line, always use all the available git options; especially when ugprading a dependency.

I have corrected the description. I just would like to update to rimage top, that is commit e9e074c3c3e8b7f97a3f07bc1eab2d2fdabcf9b6, For this reason, I added a change to the west.yml file. Is there anything else I should do? Because I don't understand the comment. e9e074c3c3e8b7f97a3f07bc1eab2d2fdabcf9b6 this is the top main in rimage.

kv2019i commented:
@fkwasowi The rimage updates are quite tricky to do. Let's do so that I'll do the rimage update in seperate PR https://github.com/thesofproject/sof/pull/6429 and you can then rebase your mux PR on top of that.

Issues that needed addressing:
 - the rimage update must be first in a series (other git bisect will be broken)
 - the commit that updates rimage commit id in west.yml, must also update the git submodule "rimage" (this is very confusing and is caused by some SOF targets not using west manifest)
 - the commit-id you put in west.yml must be a commit-id of a merge commit  (e9e074c3c3e8b7f97a3f07bc1eab2d2fdabcf9b6 is not a merged commit in rimage main branch, but rather points to prv-fkwasowi_modules_types_mtl )

fkwasowi commented:
> * git describe --dirty

I swapped the order of commits, changed the commit id to merge id, and merged two commits into one (update rimage submodules together with the change in west)-is it okay now?

fkwasowi commented:
> Thanks for updating the rimage submodule as well, however there are still many serious rimage problems:
> 
> * In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, the  commit message _subject_ says "upgrade rimage to f1e5621067bb" but the commit message _body_ says "upgrade rimage to e9e074c3c3e8b7f". At least one of them is wrong (or both)
> * In fact both are wrong because one of them is in `west.yml` but you upgraded the submodule to a third, different rimage commit d957e0368b86 instead!? As instructed in the new comments I added to `west.yml`, `git status` and `git describe --dirty` should report the submodule / west mismatch. Didn't they? Always look at "git status" before pushing (even when a single git repo is involved)
> * The latest rimage commit is currently yet another rimage commit: 3ee717eebc6a , why not upgrade to the most recent rimage commit? Is there something wrong with it? Maybe because we all love dealing with rimage upgrades so much, save another rimage upgrade for later? :-D
> * In your last force-push https://github.com/thesofproject/sof/commits/7ac73be3a070986a, your west.yml change is the LAST commit of the three. You're upgrading rimage  because your other commits need that rimage upgrade, correct? If correct then your rimage upgrade commit must be **first**. It cannot be last because every commit must be usable for `git bisect` and other test reasons.
> * Wait, actually you upgraded the rimage git submodule and west.yml in... different commits! They must be kept in sync in every commit.
> * You missed this log --oneline instruction below in the west.yml file. It's important because everyone regularly gets lost and confused with module updates - including you, see all the examples above.
> 
> ```
>   # When upgrading projects here please run git log --oneline in the
>   # project and if not too long then include the output in your commit
>   # message. See examples of good commit messages at
>   # https://github.com/thesofproject/sof/commits/4e1d3ba61abd3/rimage
>   # (git log 4e1d3ba61abd3 -- rimage/)
> ```
> 
> Make sure you find a good git plugin in your editor (or a good standalone git graphical interface), one that lets you easily review all these changes before submitting. Never restrict yourself to the limited git command line, always use all the available git options; especially when ugprading a dependency.

I swapped the order of commits, changed the commit id to merge id, and merged two commits into one (update rimage submodules together with the change in west)-is it okay now?

kv2019i commented:
Fine to me. @fkwasowi If you compare to https://github.com/thesofproject/sof/pull/6429 , the rimage history is missing from commit description, but this is not a mandatory part (and is missing from some earlier updates as well). I'm good with this version as well. We can merge this directly when Marc approves.

fkwasowi commented:
SOFCI TEST

marc-hb commented on src/audio/mux/mux.c:
```diff
@@ -34,11 +34,15 @@
-		 0xb6, 0x78, 0x7d, 0xa3, 0xc6, 0x3e, 0xa5, 0x57);
-
+		     0xb6, 0x78, 0x7d, 0xa3, 0xc6, 0x3e, 0xa5, 0x57);
+#else
```
```suggestion
#elif CONFIG_IPC_MAJOR_4
...

#else 
#error unknown IPC version
#fi
```

marc-hb reviewed (commented):
rimage commit looks good except for the lack of commit message already reported by @kv2019i . There's been a scheduled power outage this week end that affected sof-ci/jenkins and there seems to be some issues going on which leaves you plenty of time to fix the commit message.

The magic words "SOFCI TEST" only restart a couple CI checks out of many, to restart everything you must force-push. A commit date or commit message change is enough (but see my `#elif` suggestion)

fkwasowi commented:
> Fine to me. @fkwasowi If you compare to #6429 , the rimage history is missing from commit description, but this is not a mandatory part (and is missing from some earlier updates as well). I'm good with this version as well. We can merge this directly when Marc approves.

Commit message changed

lyakh commented on src/audio/copier/copier.c:
```diff
@@ -617,9 +619,10 @@ static int copier_prepare(struct comp_dev *dev)
+		for (int i = 0; i < IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT; i++)
+			cd->converter[i] = get_converter_func(&cd->config.base.audio_fmt,
+							      &cd->config.out_fmt, ipc4_gtw_none,
+							      ipc4_bidirection);
```
`get_converter_func()`'s parameters don't change with the loop iterations, so, the result will always be the same? Then maybe better keep the above line for `cd->converter[0]` then the below check and after that in an `else` branch copy from index 0 to all other array members?

fkwasowi replied:
Fixed

marc-hb reviewed (changes requested):
Fails to compile TGL in https://sof-ci.01.org/sof-pr-viewer/#/build/PR6385/build10613048


```
/localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc -DBLD_COUNTERS=0 -DRELATIVE_FILE=\"src/audio/copier/copier.c\" -I/quickbuild/workspace1/24733/SOF_FW/src/platform/intel/cavs/include -I/quickbuild/workspace1/24733/SOF_FW/src/arch/xtensa/include -I/quickbuild/workspace1/24733/SOF_FW/src/arch/xtensa/xtos -I/quickbuild/workspace1/24733/SOF_FW/xtos/include -I/quickbuild/workspace1/24733/SOF_FW/src/platform/tigerlake/include -I/localdisk/tools/xtensa/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8/xtensa-elf/arch/include -I/localdisk/tools/xtensa/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8/xtensa-elf/include -I/quickbuild/workspace1/24733/SOF_FW/rimage/src/include -I/quickbuild/workspace1/24733/SOF_FW/src/include -I/quickbuild/workspace1/24733/SOF_FW/build_fw/generated/include -I/quickbuild/workspace1/24733/SOF_FW/third_party_include -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacros/quickbuild/workspace1/24733/SOF_FW/build_fw/generated/include/autoconfig.h -MD -MT CMakeFiles/sof.dir/src/audio/copier/copier.c.o -MF CMakeFiles/sof.dir/src/audio/copier/copier.c.o.d -o CMakeFiles/sof.dir/src/audio/copier/copier.c.o -c /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c
/quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c: In function ‘copier_new’:
/quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c:463: error: ‘for’ loop initial declaration used outside C99 mode
/quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c: In function ‘copier_prepare’:
/quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c:622: error: ‘for’ loop initial declaration used outside C99 mode
CMakeFiles/sof.dir/build.make:1037: recipe for target 'CMakeFiles/sof.dir/src/audio/copier/copier.c.o' failed
make[3]: *** [CMakeFiles/sof.dir/src/audio/copier/copier.c.o] Error 2
```

TGL compiles just fine in https://sof-ci.01.org/sofpr/PR6385/build1949/build/tgl_zph.txt with `xcc/install/builds/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8`, in my workspace and (hopefully) in your workspace too? What toolchain did you use?

So this looks a CI misconfiguration problem. CI should never compile anything with "secret flags" or options that developers cannot reproduce easily, all the entire build configurations MUST be available in version control. No CI "secret sauce" ever.

Whatever the root cause is, we can't merge code that does not compile for whatever reason.

fkwasowi commented:
> Fails to compile TGL in https://sof-ci.01.org/sof-pr-viewer/#/build/PR6385/build10613048
> 
> ```
> /localdisk/tools/xtensa/RG-2017.8-linux/XtensaTools/bin/xt-xcc -DBLD_COUNTERS=0 -DRELATIVE_FILE=\"src/audio/copier/copier.c\" -I/quickbuild/workspace1/24733/SOF_FW/src/platform/intel/cavs/include -I/quickbuild/workspace1/24733/SOF_FW/src/arch/xtensa/include -I/quickbuild/workspace1/24733/SOF_FW/src/arch/xtensa/xtos -I/quickbuild/workspace1/24733/SOF_FW/xtos/include -I/quickbuild/workspace1/24733/SOF_FW/src/platform/tigerlake/include -I/localdisk/tools/xtensa/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8/xtensa-elf/arch/include -I/localdisk/tools/xtensa/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8/xtensa-elf/include -I/quickbuild/workspace1/24733/SOF_FW/rimage/src/include -I/quickbuild/workspace1/24733/SOF_FW/src/include -I/quickbuild/workspace1/24733/SOF_FW/build_fw/generated/include -I/quickbuild/workspace1/24733/SOF_FW/third_party_include -nostdlib -fno-inline-functions -mlongcalls -O2 -g -Wall -Werror -Wl,-EL -Wmissing-prototypes -Wpointer-arith -mtext-section-literals -imacros/quickbuild/workspace1/24733/SOF_FW/build_fw/generated/include/autoconfig.h -MD -MT CMakeFiles/sof.dir/src/audio/copier/copier.c.o -MF CMakeFiles/sof.dir/src/audio/copier/copier.c.o.d -o CMakeFiles/sof.dir/src/audio/copier/copier.c.o -c /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c
> /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c: In function ‘copier_new’:
> /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c:463: error: ‘for’ loop initial declaration used outside C99 mode
> /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c: In function ‘copier_prepare’:
> /quickbuild/workspace1/24733/SOF_FW/src/audio/copier/copier.c:622: error: ‘for’ loop initial declaration used outside C99 mode
> CMakeFiles/sof.dir/build.make:1037: recipe for target 'CMakeFiles/sof.dir/src/audio/copier/copier.c.o' failed
> make[3]: *** [CMakeFiles/sof.dir/src/audio/copier/copier.c.o] Error 2
> ```
> 
> TGL compiles just fine in https://sof-ci.01.org/sofpr/PR6385/build1949/build/tgl_zph.txt with `xcc/install/builds/RG-2017.8-linux/cavs2x_LX6HiFi3_2017_8` and (hopefully) in your workspace too? What toolchain did you use?
> 
> So this looks a CI misconfiguration problem. CI should never compile anything with "secret flags" or secret options / configurations that developers cannot reproduce easily, all the entire build configurations MUST be available in version control. No CI "secret sauce" ever.
> 
> Whatever the root cause is, we can't merge code that does not compile for whatever reason.

Fixed

marc-hb commented:
> So this looks a CI misconfiguration problem. CI should never compile anything with "secret flags" or secret options / configurations that developers cannot reproduce easily, all the entire build configurations MUST be available in version control. No CI "secret sauce" ever.

> Fixed

What was fixed, is the CI secret sauce gone? It looks like you just changed the code to avoid the issue.

marc-hb commented:
> Fails to compile TGL in https://sof-ci.01.org/sof-pr-viewer/#/build/PR6385/build10613048

Apologies, this was not a Zephyr build. Thanks @mwasko for the tip.

aborisovich commented:
SOF CI

fkwasowi commented:
@kv2019i , You recently released a change with the same failing step (https://github.com/thesofproject/sof/pull/6383), "Known fail in IPC4 test set, proceeding with merge."-could I also ask for such a merge? Because I do not know how to pass the step sof-ci. I have the same problem as Tomasz Lissowski had.

marc-hb commented:
I merged @kv2019i 's hard work in https://github.com/thesofproject/sof-test/pull/971 so https://sof-ci.01.org/sofpr/PR6385/build2000/devicetest is all green. No need to ask for any exception now
